orbit/pkg/update/notifications.go

```diff
@@ -0,0 +1,76 @@
+package update
+
+import (
+	"sync"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/service"
+	"github.com/rs/zerolog/log"
+)
+
+type runCmdFunc func() error
+
+// RenewEnrollmentProfileConfigFetcher is a kind of middleware that wraps an
+// OrbitConfigFetcher and detects if the fleet server sent a notification to
+// renew the enrollment profile. If so, it runs the command (as root) to
+// bootstrap the renewal of the profile on the device (the user still needs to
+// execute some manual steps to accept the new profile).
+//
+// It ensures only one renewal command is executed at any given time, and that
+// it doesn't re-execute the command until a certain amount of time has passed.
+type RenewEnrollmentProfileConfigFetcher struct {
+	// Fetcher is the OrbitConfigFetcher that will be wrapped. It is responsible
+	// for actually returning the orbit configuration or an error.
+	Fetcher OrbitConfigFetcher
+	// Frequency is the minimum amount of time that must pass between two
+	// executions of the profile renewal command.
+	Frequency time.Duration
+
+	// for tests, to be able to mock command execution. If nil, will use
+	// runRenewEnrollmentProfile.
+	runCmdFn runCmdFunc
+
+	// ensures only one command runs at a time, protects access to lastRun
+	cmdMu   sync.Mutex
+	lastRun time.Time
+}
+
+// GetConfig calls the wrapped Fetcher's GetConfig method, and if the fleet
+// server set the renew enrollment profile flag to true, executes the command
+// to renew the enrollment profile.
+func (h *RenewEnrollmentProfileConfigFetcher) GetConfig() (*service.OrbitConfig, error) {
+	cfg, err := h.Fetcher.GetConfig()
+	if err == nil && cfg.Notifications.RenewEnrollmentProfile {
+		if h.cmdMu.TryLock() {
+			defer h.cmdMu.Unlock()
+
+			// TODO: what's a good delay after the last "renew enrollment profile"
+			// command has run where we can assume Fleet has been notified that this
+			// host is enrolled? AFAICS there's still a manual step that the user
+			// must execute after the renew command has been executed - they will
+			// receive a pop-up notification and then must follow some steps to
+			// accept the new enrollment profile, captured in this figma:
+			// https://www.figma.com/file/hdALBDsrti77QuDNSzLdkx/%F0%9F%9A%A7-Fleet-EE-(dev-ready%2C-scratchpad)?node-id=10566%3A315998&t=jSV3l5p3I5gxX4EW-0
+			//
+			// So the host could be actually enrolled only much later, resulting in
+			// this command being executed multiple times (presumably with the
+			// associated macOS notification popup). Is that ok? Is that command
+			// idempotent apart from the notification popup?
+			if time.Since(h.lastRun) > h.Frequency {
```

Comment:
@mna 

> is it safe to continually execute the profiles renew command until the user completes the manual steps and the Fleet server gets notified that the device is now enrolled, and stops sending the notification flag to true?

It is safe in the sense that nothing can go wrong for us, it could be annoying/confusing for the user as you suggest.

I have recorded here different scenarios and how the notification popup looks like:

https://user-images.githubusercontent.com/4419992/214114075-22592607-3bf3-4273-856c-c888a744ea2b.mov

In summary:

- If the pop-up is closed, it will be opened again
- If the pop-up is already open, nothing visibly happens
- If the user already completed the enrollment, the popup still appears again

IMO, we should increase the value of `h.Frequency` to something like 1 hour, we still want to be a bit annoying (the sooner the user enrolls, the better) but 5 minutes might be too low. What do you think? should we ask Noah?

> When the user completes the manual steps, presumably by accepting the new profile, it will trigger the MDM checkin message ("Authenticate"), which should update the enrolled status of the device. Or would it? Looking at https://github.com/fleetdm/fleet/blob/main/server/datastore/mysql/apple_mdm.go#L267-L282, if the host already exists in our tables (which would be the case if the host was migrated from another mdm solution, but was otherwise already enrolled in fleet), nothing is done. Is the idea here that we will update the host's mdm enrollment information only once we receive the information via the osquery MDM detail query?

All what you said here is correct, but I think we can remove that `foundHost.HardwareSerial != mdmHost.SerialNumber || foundHost.UUID != mdmHost.UDID`, IIRC, @gillespi314 added it as an optimization.

Even if we increase`h.Frequency` to be higher, we can't completely depend on osquery as the interval for detail queries (where we get the MDM information) can be configured by the user.

Another option is to set `host.refetch_requested`, but it's still not 100% bulletproof since the check-in frequency can also be configured!

Reply:
Thanks for the detailed investigation, @roperzh , great job! I'll ping @noahtalerman as I think it's important he's aware of the behaviour (Noah: I think Roberto's video covers very well the various ways this could behave).

In light of what you've found, I agree that making that every hour makes more sense. I'll update the PR.

> I think we can remove that foundHost.HardwareSerial != mdmHost.SerialNumber || foundHost.UUID != mdmHost.UDID, IIRC, @gillespi314 added it as an optimization.

Note that even if we remove this condition, the code would still not update the `host_mdm` table (the `updateMDMAppleHostDB` function only updates the main `hosts` table), we'd still have to wait until osquery reports the new MDM enrollment status before the notification stops showing up periodically. If that's not acceptable, then I believe we'd have to add a call to `upsertMDMAppleHostMDMInfoDB` from the `updateMDMAppleHostDB` function _in addition_ to removing the condition.

Reply:
Hey! Thanks for creating a video. I haven't had a chance to check this out. Planning on watching Roberto's video today.